src/server/db_slice.cc

```diff
@@ -364,86 +366,122 @@ DbSlice::AddOrFindResult& DbSlice::AddOrFindResult::operator=(ItAndUpdater&& o)
+  return {res->it, res->exp_it};
 }
 
 OpResult<PrimeConstIterator> DbSlice::FindReadOnly(const Context& cntx, string_view key,
                                                    unsigned req_obj_type) {
-  auto it = FindReadOnly(cntx, key).it;
-
-  if (!IsValid(it))
-    return OpStatus::KEY_NOTFOUND;
-
-  if (it->second.ObjType() != req_obj_type) {
-    return OpStatus::WRONG_TYPE;
+  auto res = FindInternal(cntx, key, req_obj_type, UpdateStatsMode::kUpdateReadStats,
+                          LoadExternalMode::kDontLoadExternal);
+  if (res.ok()) {
+    return {res->it};
   }
-
-  return {it};
+  return res.status();
 }
 
-DbSlice::ItAndExp DbSlice::FindInternal(const Context& cntx, std::string_view key,
-                                        FindInternalMode mode) {
-  DbSlice::ItAndExp res;
+OpResult<PrimeConstIterator> DbSlice::FindAndFetchReadOnly(const Context& cntx,
+                                                           std::string_view key,
+                                                           unsigned req_obj_type) {
+  auto res = FindInternal(cntx, key, req_obj_type, UpdateStatsMode::kUpdateReadStats,
+                          LoadExternalMode::kLoadExternal);
+  if (res.ok()) {
+    return {res->it};
+  }
+  return res.status();
+}
 
-  if (!IsDbValid(cntx.db_index))
-    return res;
+OpResult<DbSlice::ItAndExp> DbSlice::FindInternal(const Context& cntx, std::string_view key,
+                                                  std::optional<unsigned> req_obj_type,
+                                                  UpdateStatsMode stats_mode,
+                                                  LoadExternalMode load_mode) {
+  DCHECK(IsDbValid(cntx.db_index));
 
+  DbSlice::ItAndExp res;
   auto& db = *db_arr_[cntx.db_index];
   res.it = db.prime.Find(key);
-  FiberAtomicGuard fg;
-  if (!IsValid(res.it)) {
-    switch (mode) {
-      case FindInternalMode::kUpdateMutableStats:
+
+  auto update_stats_on_miss = [&](UpdateStatsMode stats_mode) {
+    switch (stats_mode) {
+      case UpdateStatsMode::kUpdateMutableStats:
         events_.mutations++;
         break;
-      case FindInternalMode::kUpdateReadStats:
+      case UpdateStatsMode::kUpdateReadStats:
         events_.misses++;
         break;
     }
-    return res;
+  };
+
+  if (!IsValid(res.it)) {
+    update_stats_on_miss(stats_mode);
+    return OpStatus::KEY_NOTFOUND;
   }
 
+  if (req_obj_type.has_value() && res.it->second.ObjType() != req_obj_type.value()) {
+    return OpStatus::WRONG_TYPE;
+  }
+
+  if (TieredStorage* tiered = shard_owner()->tiered_storage();
+      tiered && load_mode == LoadExternalMode::kLoadExternal) {
+    if (res.it->second.HasIoPending()) {
+      tiered->CancelIo(cntx.db_index, res.it);
+    } else if (res.it->second.IsExternal()) {
+      // Load reads data from disk therefore we will preempt in this function.
+      // We will update the iterator if it changed during the preemption
```

Comment:
What about the rear case in which we'll receive a `FLUSHALL` (or `FLUSHDB`) while preempted?
That should invalidate `db`, so we may need to use a pointer here and update that as well?

Reply:
I believe we will not be able to invoke FLUSHALL while we are reading , as read will acquier a read lock and the FLUSHALL will do a write lock on all shards and it will not execute untill we reach it in the tx queue